### PR TITLE
[Model Averaging] Create a base class for model averaging

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62132
* #62131
* __->__ #62111

This base class will be passed to the post-localSGD optimizer in the next PR. This way, the same post-localSGD optimizer can choose different model averaging algorithms.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29884954](https://our.internmc.facebook.com/intern/diff/D29884954/)

### CODE REVIEW

File: torch/distributed/algorithms/model_averaging/averagers.py [hunk 1]
```diff
@@ -4,7 +4,29 @@
...
+    """
+
+    def __init__(self, params, process_group=None):
+        self.params = list(params)
+        self.process_group = (
+            process_group if process_group is not None else dist.group.WORLD
+        )
+        self.step = 0
+
+    def average_parameters(self):
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/averagers.py:25]: Should we consider making `ModelAverager` an abstract class and enforcing that `average_parameters` is overridden, or at least raise a NotImplementedError here?

**Review by rohan-varma — APPROVED** [review]
lgtm, one nit

File: torch/distributed/algorithms/model_averaging/averagers.py [hunk 2]
```diff
@@ -4,7 +4,29 @@
...
+        pass
+
+
+class PeriodicModelAverager(ModelAverager):
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/averagers.py:29]: Do we plan to add additional model averaging impls in the future? If so, do we have an idea what those could be?

Comment (SciPioneer) [reply torch/distributed/algorithms/model_averaging/averagers.py:29]: I do have prototyped another more complicated implementation, and SlowMo can be another candidate here. The point is that when I wrap this kind of model averaging component in an optimizer, I want to use an interface as the typing info.

**Pull request closed without merging**